src/components/controls/controls.js

```diff
@@ -0,0 +1,225 @@
+var registerControls = require('../../core/controls').registerControls;
+var THREE = require('../../lib/three');
+
+var MAX_DELTA = 0.2; // ms
+var PI_2 = Math.PI / 2;
+
+/**
+ * Controls component.
+ *
+ * Receives input events from device-specific control components, and applies movement and rotation
+ * to the element accordingly.
+ */
+module.exports.Component = registerControls('controls', {
+  dependencies: ['velocity'],
+
+  schema: {
+    enabled: { default: true },
+
+    flyingEnabled: { default: false },
+
+    position: {
+      default: ['hmd-controls', 'gamepad-controls', 'touch-controls']
+    },
+    positionControlsEnabled: { default: true },
+    positionEasing: { default: 15 }, // m/s2
+    positionAcceleration: { default: 65 }, // m/s2
+
+    rotation: { default: ['hmd-controls', 'gamepad-controls', 'mouse-controls'] },
+    rotationEnabled: { default: true }
+  },
+
+  init: function () {
+    // Movement
+    this.velocity = new THREE.Vector3();
+
+    // Rotation
+    this.pitch = new THREE.Object3D();
+    this.yaw = new THREE.Object3D();
+    this.yaw.position.y = 10;
+    this.yaw.add(this.pitch);
+    this.heading = new THREE.Euler(0, 0, 0, 'YXZ');
+
+    if (this.el.sceneEl.hasLoaded) {
+      this.validateControls();
+    } else {
+      this.el.sceneEl.addEventListener('loaded', this.validateControls.bind(this));
+    }
+  },
+
+  update: function (previousData) {
+    // Re-validate if scene has loaded. If not, listener was already bound in init().
+    if (!this.el.sceneEl.hasLoaded) { return; }
+    this.validateControls();
+  },
+
+  tick: function (t, dt) {
+    if (isNaN(dt)) { return; }
+
+    // Update rotation.
+    if (this.data.rotationEnabled) {
+      this.updateRotation(dt);
+    }
+
+    // Update velocity. If FPS is too low, reset.
+    if (this.data.positionControlsEnabled && dt / 1000 > MAX_DELTA) {
+      this.velocity.set(0, 0, 0);
+      this.el.setAttribute('velocity', this.velocity);
+    } else if (this.data.positionControlsEnabled) {
+      this.updateVelocity(dt);
+    }
+  },
+
+  /**
+   * Updates rotation based on input from the active rotation component, if any.
+   * @param  {number} dt
+   */
+  updateRotation: function (dt) {
+    var dRotation;
+    var el = this.el;
+    var yaw = this.yaw;
+    var pitch = this.pitch;
+    var control = this.getActiveRotationControls();
+    if (control && control.getRotationDelta) {
+      dRotation = control.getRotationDelta(dt);
+      yaw.rotation.y -= dRotation.x;
+      pitch.rotation.x -= dRotation.y;
+      pitch.rotation.x = Math.max(-PI_2, Math.min(PI_2, pitch.rotation.x));
+      el.setAttribute('rotation', {
+        x: THREE.Math.radToDeg(pitch.rotation.x),
+        y: THREE.Math.radToDeg(yaw.rotation.y),
+        z: 0
+      });
+    } else if (control) {
+      el.setAttribute('rotation', control.getRotation());
```

Comment:
Pitch/yaw do get stored, I think maybe what you're seeing is that we don't work backwards to compute pitch and yaw from HMD `getRotation` updates?

Off hand I would think we need:

```javascript
yaw.rotation.y = THREE.Math.degToRad(rotation.y);
pitch.rotation.x = THREE.Math.degToRad(rotation.x);
pitch.rotation.x = Math.max(-PI_2, Math.min(PI_2, pitch.rotation.x));
```